src/workerd/io/actor-sqlite.c++

```diff
@@ -513,10 +523,21 @@ ActorCacheInterface::DeleteAllResults ActorSqlite::deleteAll(WriteOptions option
     }
   }
 
-  if (localAlarmState == kj::none && !deleteAllCommitScheduled) {
-    // If we're not going to perform a write to restore alarm state, we'll want to make sure the
-    // commit callback is called for the deleteAll().
-    commitTasks.add(outputGate.lockWhile(kj::evalLater([this]() mutable -> kj::Promise<void> {
+  // We want to ensure a commit is scheduled during the deleteAll() call, but if we end up doing a
+  // write to restore alarm state afterward, we can skip the original scheduled commit, since the
+  // write will queue its own commit.
+  kj::Maybe<kj::Own<bool>> maybeCanSkipDeleteAllCommit;
+
+  if (!deleteAllCommitScheduled) {
```

Comment:
In case it isn't apparent: the main fix here is removal of the `localAlarmState == kj::none` check, to ensure `deleteAllCommitScheduled` is always true by the time we reach `db.clear()`, which in turn ensures that `isCommitScheduled()` is true, since it's used to check an invariant in a storage relay callback.

The `localAlarmState == kj::none` check was originally put in place to avoid having a redundant commit, since the write to restore the alarmState also adds a commit.  But of course there's no way for the storage relay callback to know via the current `isCommitScheduled()` implementation that we're planning to add the commit after calling `db.clear()` (and maybe it isn't even correct to rely solely on committing after `db.clear()`, if it's important that the output lock is synchronously started before calling `db.clear()`?).

Reply:
Hmm, I think you're overestimating the harm of a "redundant commit". `StorageRelayClient::commit()` already detects if there's nothing to commit (`if (!needCommit` at the beginning), so it's not that big a deal to schedule an extra one here.

And `deleteAll()` is not commonly called so I don't think it's worth optimizing away some promises.

Let's instead simplify this code to just always schedule (and apply) the deleteAll commit.

Reply:
- [ ] Sure, makes sense... I'll try to simplify it.

Reply:
I guess what you're getting at is that the reset commit task doesn't actually do the same set of work as a regular write commit -- it doesn't actually commit a transaction and whatnot. Yeah, I agree it would be a problem if we committed the reset without actually committing the transaction to restore the alarm time.

So perhaps the reset commit task needs to actually skip calling `commitCallback` in the case that it knows that a further commit is coming for the metadata. Perhaps it can just be like:

```
    commitTasks.add(outputGate.lockWhile(kj::evalLater([this]() mutable -> kj::Promise<void> {
      // Don't commit if shutdown() has been called.
      requireNotBroken();

      deleteAllCommitScheduled = false;
      if (currentTxn.is<ImplicitTxn>()) {
        // An implicit transaction is already scheduled, so we'll count on it to perform a commit when it's
        // done. This is particularly important for the case where deleteAll() was called while an alarm
        // is outstanding; we'll immediately begin an implicit transaction that restores the alarm. We don't
        // want to commit the deletion without that transaction.
        return kj::READY_NOW;
      } else {
        return commitCallback();
      }
    })));
    deleteAllCommitScheduled = true;
```

Reply:
Another, maybe cleaner approach would be to unify the two commit tasks into a single one which checks both `currentTxn.is<ImplicitTxn>()` and `deleteAllCommitScheduled` to decide what actual work needs to be performed before committing?

Reply:
I was also curious... I assume it's OK/moot that we call `requireNotBroken()` prior to clearing `deleteAllCommitScheduled`, since once the output gate is broken, we expect all subsequent user operations to fail and continue failing, such that the return value of `isCommitScheduled()` no longer matters?

Reply:
Yeah nothing matters once broken.

Reply:
I tried combining the `deleteAll()` and `onWrite()` commit tasks, but it was turning out to be a more complicated change than your suggestion of just checking `if (currentTxn.is<ImplicitTxn*>()) {...}` in `deleteAll()`, so I went with that.

One thing I wasn't quite sure of, though, is the case involving explicit transactions:

- user calls `deleteAll()` while alarm state is present.
- `deleteAll()` adds its `kj::evalLater()` commit task to the event queue.
- `deleteAll()` calls `setAlarm()` to restore the post-delete alarm state, which fires `onWrite()`, which sets up the implicit transaction and adds its `kj::evalLater()` commit task to the event queue to run immediately after.
- before the event queue processes either commit task, it runs an event where user code starts an explicit transaction.
- the explicit transaction constructor runs the implicit transaction's commit(), then updates `currentTxn` to point to itself.
- the user code maybe makes a few additional changes within the explicit transaction.

When the `deleteAll()` commit task runs, it will see that `currentTxn` is no longer an implicit transaction, so it's going to just run `commitCallback()`.  Is that OK?

I assume the `commitCallback()` would end up persisting the interim state of the explicit transaction.  But presumably that's OK, because if the system goes down at that point, we recover from that by rolling back any explicit transaction savepoints we see in the persisted DB when reloading?  (If not, I think a similar sequence of events is already possible just using implicit and explicit transactions together, without calling `deleteAll()`.)

And I guess in that case, we'd be running a `commitCallback()` that effectively updates alarm state without also running the alarm rescheduling commit code, but maybe that's OK, since the very next event queue item should be the implicit transaction's commit task, which would run it?